### PR TITLE
[Modular] Ports Integrated Electronics (custom circuitry)

### PR DESCRIPTION
**Pull request by NotRanged**
## About The Pull Request

After many hours of fixing hundreds of compile and logic errors, it is finally ready.
This PR ports the latest version of Integrated Electronics/circuitry from Beestation, along with various features from Polaris' version of it (most notably, the clothing.)

Until this PR is fullmerged (test-merge only for a while is advised) there won't be a circuit-lab mapped into science on the stations yet. The default printer and clothing will all be printable by the science protolathe and start fully upgraded. (No upgrade disks, this can be added later).

I'm expecting a few strange problems and runtimes still since this is too big of a feature for me to test fully by myself, so please put all problems regarding circuitry in this PR or DM them to me on Discord (Ranged#0001). The basics of it work fine in testing though.

Currently, some of the NTNet circuits are commented out because tg fully changed how NTNet works, I'll work on them later. Reagent circuits can't inject into hydro trays for similar reasons. Weapons mechanism has been commented out since it was a source of problems on the old codebase.

_Tests-The-Changes in full circuitry gear._
image

_Printer and UI_
image

## Why It's Good For The Game

The latest Beestation version of a feature that many enjoyed on the old codebase is now ported. More stuff for scientists to do.

## Changelog
:cl:
add: Ported the latest version of Integrated Electronics (Circuitry) from Beestation.
add: You can print the Circuit Printer from the Protolathe in science.
add: Ported Intergrated Electronic Clothing from Polaris. These have action buttons you can press to activate their internal custom circuits. They can be printed from the Protolathe in science.
/:cl:

### CODE REVIEW

**Comment by NotRanged** [comment]
Oh boy here come the 100 linter errors

**Comment by Arctaisia** [comment]
My hero :heart:

**Comment by SarmentiCampbell** [comment]
that awesome.

**Comment by YakumoChen** [comment]
Ahhh, TG itself has tried this before and it came with the burden of unmaintainable code, dozens of exploits, and was generally a pain so they removed it.

Do you really think this is a good idea?

**Comment by NotRanged** [comment]
> Ahhh, TG itself has tried this before and it came with the burden of unmaintainable code, dozens of exploits, and was generally a pain so they removed it.
> 
> Do you really think this is a good idea?

latest beestation version, not cit, also yes

**Comment by SpaceVampire** [comment]
I don't want to put this down, because of the work you put into it.
But part of does agree with Chen, that circuits was removed for a good reason.
That it tends to be super heavy when it comes to exploits, and just being a massive pain in general.

**Comment by Arctaisia** [comment]
> I don't want to put this down, because of the work you put into it.
> But part of does agree with Chen, that circuits was removed for a good reason.
> That it tends to be super heavy when it comes to exploits, and just being a massive pain in general.

That's why Ranged made sure that exploits are harder to come by, of not impossible, and with the removal of weapon circuits, there's less gatling guns. Bee's port is pretty solid, plus, there's many more positives to circuits than negative, and a scientist that knows what they're doing can help the station immensely woth circuot designs.

Another part, if you see anyone exploiting circuits, report it, removing a feature that's used by 10s of people, because one was a shitter, makes it unfun for all those others.

**Comment by YakumoChen** [comment]
Some questions:
-Does this have logging 
-If you examine a circuit bot will you be able to tell what it's doing
--What parts are on it
--and what items are attached to it or being held (like guns etc)
-is there a UI

These are important questions and, keep in mind, this is a 10,000 line PR that maintainers will have to go through with a fine tooth comb to make sure it "works".

**Comment by NotRanged** [comment]
> is there logging

Yes
 > is there a UI

Literally in the PR body...

> what parts are on it

if you open it up you can tell

Once again, it's the latest beestation version of it, and they're known for keeping their code as exploit free to pretty much the same standard as tg.

> this is a 10,000 line PR that maintainers will have to go through with a fine tooth comb to make sure it "works".

Not really how it works

**Comment by unit0016** [comment]
Fuck yes.
I'd like to note that the alien multitool doesn't have all the icon states necessary for circuitry yet though, weh.

**Comment by Cheshify** [comment]
Absolute madman did it

**Comment by Draganfrukts** [comment]
The bobmed equivalent of newbase

**Comment by Gandalf2k15** [comment]
Alright I think we can give this a test merge. Let's see how badly it affects performance.

**Comment by Gandalf2k15** [comment]
After conflicts are fixed

**Comment by Gandalf2k15** [comment]
Ah, you'll have to update pathfinding to use the new JPS system, see https://github.com/Skyrat-SS13/Skyrat-tg/pull/3581

**Comment by unit0016** [comment]
image
???

**Comment by unit0016** [comment]
Video camera circuit bugged.

**Comment by unit0016** [comment]
Boots are invisible.
Jumpsuits don't display right.

File: modular_skyrat/modules/integrated_circuits/code/integrated_electronics/subtypes/output.dm [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+	.= ..()
+	sounds = GLOB.vox_sounds
+	extended_desc = "The first input pin determines which sound is used. It uses the AI Vox Broadcast word list. So either experiment to find words that work, or ask the AI to help in figuring them out. The second pin determines the volume of sound that is played, and the third determines if the frequency of the sound will vary with each activation."
+
+/obj/item/integrated_circuit/output/text_to_speech
+	name = "text-to-speech circuit"
+	desc = "Takes any string as an input and will make the device say the string when pulsed."
+	extended_desc = "This unit is more advanced than the plain speaker circuit, able to transpose any valid text to speech."
+	icon_state = "speaker"
+	cooldown_per_use = 10
```
Comment (Elli-Skala) [inline modular_skyrat/modules/integrated_circuits/code/integrated_electronics/subtypes/output.dm:232]: after messing around with circits id sudgest moving this down from 1 second to 0.1 second. as its mostly used for local translator. while chat spam may seem like a issue text to radio and lets everyone hear already sends the messages as up to 0.1

**Comment by Elli-Skala** [comment]
seems to have promises but a few issues I've noticed is that circuit cloning doesn't work, and on old base all the parts use to be made of iron letting you recycler them back into the machine if you didn't want to use them. you cant do that at all. also seems the id lock on the circuit printer doesn't work?

**Comment by unit0016** [comment]
Please let ancient lathes make this stuff.

**Review by Gandalf2k15 — APPROVED** [review]

**Comment by Gandalf2k15** [comment]
Test merge period is complete. Full merge approval granted.

**Pull request merged**